### ISSUE TITLE
Useful CUDA libraries

### ISSUE THREAD

**Issue by m-pilia** [simeks/stk#8]
Hi, in the last few days I ported disptools to CUDA. Since disptools is a tiny library I implemented my kernels from scratch, but while doing that I was thinking about possible support libraries that may be useful for a larger project like stk/deform.

I think Thrust is very nice and handy, with a lot of high-level convenience features, but it is rather slow and it requires to use its own object types, so maybe we should avoid it. CUB on the other hand is a lower level library that offers a lot of interesting routines, it is much faster than Thrust, and it allows to operate directly on raw pointers, so it may be useful for us, so we do not have to reinvent the wheel every time.

One thing I tried to implement is a [generic reduction routine](https://github.com/m-pilia/disptools/blob/master/cuda/headers/reduction.cuh) that supports incompatible input and output types and allows a transform from the input type to the output type within the reduction itself (e.g. to simultaneously compute minimum and maximum of an array, or to compute the squared norm of a large array). I am not sure, but I do not think this is available in CUB, which has some [interesting reductions](https://nvlabs.github.io/cub/structcub_1_1_device_reduce.html) nevertheless. It should be possible in Thrust, combining a few primitives, but I bet the result would be slow. In case we find the need for something like it, maybe we can clean up and reuse my code in stk/deform as well.

The one thing I wanted but I did not manage to achieve so far is to remove the `__syncthreads()` from the last warp: the warp is already synchronous, but on modern architectures that requires to mark the memory as `volatile` to ensure that the result is stored before the next read. That's easy to do with something like
```cuda
template <typename T, typename T_op>
__device__ __forceinline__ void reduce_warp(volatile T *sdata, const T_op op, const int tid) {
    sdata[tid] = op(sdata[tid], sdata[tid + 32]);
    sdata[tid] = op(sdata[tid], sdata[tid + 16]);
    sdata[tid] = op(sdata[tid], sdata[tid + 8]);
    sdata[tid] = op(sdata[tid], sdata[tid + 4]);
    sdata[tid] = op(sdata[tid], sdata[tid + 2]);
    sdata[tid] = op(sdata[tid], sdata[tid + 1]);
}
```
to be called inside `if (tid < 32)` in the end of the reduction, which works well for scalar types, but I had trobles if `T` is a vector (like `float3` or similar), since it seems there is no implicit conversion between `volatile` and non `volatile` vector types (which breaks `op`, and even the assignment operator).

**Comment by simeks** [comment]
I've been looking at a couple of tools but my impression is that everyone hates 3D data, haha. The fact that I haven't really decided between texture and pitched memory doesn't make it easier. I ended up doing everything on my own mostly because it wasn't that much work but I guess at least STK could benefit by having some generic functions.

I wrote the min/max reduction for STK with the thought of making it generic but after a while I just did bare minimum to get it to work. I got at bit disappointed by the fact that [this](https://devblogs.nvidia.com/cooperative-groups/) didn't work on Windows, because that would make for a really nice single-pass kernel.

Yeah, that one is tricky. They say that all threads in a warp should be synchronous so `__syncthreads()` isn't needed, but they still do it in the samples. 

Here's a snippet from your `reduction.cuh`:

```
if (num_threads >= 64) {
    s_data[tid] = op(s_data[tid], s_data[tid + 32]);
}
if (num_threads >= 32) {
    s_data[tid] = op(s_data[tid], s_data[tid + 16]);
}
....
```

I think there's an error in there.  The second assignment should only pass for `(tid < 16 && num_threads >= 32)`. If i didn't miss anything you only check for `tid < 32`. So you're reading out of bounds.

**Comment by m-pilia** [comment]
Yep, that's a problem with compiler optimisation. The threads within the warp are synchronous, but the compiler may decide to optimise and keep the result within a register without storing it between the statements, being unaware that the location of the store will be read by other threads in the next statement. Unless it is forced to do so, due to `__syncthreads()` or having the memory qualified as `volatile`. Now, `__syncthreads()` introduces unnecessary overhead, so it would be nice to just force the stores with `volatile`, but for non-scalar types apparently that messes with the type system for some reason...
I saw some old examples without `__syncthreads()` nor `volatile`, but I think that is because pre-Volta GPUs had separate instructions to operate on shared memory that did not involve regular loads and stores, so they were not subject to this problem.

For my code, I do not feel super confident, but I am not sure I see the problem (I may be wrong though). If I understand correctly what you mean, you say that I should check for `tid < 16` within the macro at line 77 (and similar for the following lines). Now for convenience I copy here the snippet at lines 44-47
```cuda
REDUCE_SYNC(64, op)
if (tid < 32) {
    REDUCE_SIMD(32, op)
    REDUCE_SIMD(16, op)
```
which expands to
```cuda
if (num_threads >= 2 * 64) {
    if (tid < 64) {
        s_data[tid] = op(s_data[tid], s_data[tid + 64]);
    }
    __syncthreads();
}
if (tid < 32) {
    if (num_threads >= 2 * 32) {
        s_data[tid] = op(s_data[tid], s_data[tid + 32]);
        __syncthreads();
    }
    if (num_threads >= 2 * 16) { // not checking for tid < 16 !!!
        s_data[tid] = op(s_data[tid], s_data[tid + 16]);
        __syncthreads();
    }
    ...
```
Within the last warp I do not check the `tid` anymore, since there is only one warp active indeed, so masking out the threads with higher `tid` slows the flow but does not save any work. The values stored by those threads are garbage, but they are not read again by anyone, so it should not matter. There should be no read out of bound (I debugged the code with `cuda-memcheck` earlier today) since due to `num_threads >= 2 * n_` I am always checking that the highest address that is read read falls within the block size (hence within the size of the shared memory currently allocated).

**Comment by m-pilia** [comment]
For cooperative groups, yeah, that would be cool, since with the current approach block synchronisation is done with host code... Even though I must admit I am not familiar with them, it is pretty recent stuff I think.

**Comment by simeks** [comment]
What types do you have problems with? Could be an alignment issue.

Say you have 32 threads, then you'll want to reduce these 32 values to 16 by

```
s[0] = reduce(s[0], s[0+16])
s[1] = reduce(s[1], s[1+16])
...
s[15] = reduce(s[15], s[15+16])
```

You'll only want to process half the range, but what you're doing now is:

```
s[0] = reduce(s[0], s[0+16])
s[1] = reduce(s[1], s[1+16])
...
s[15] = reduce(s[15], s[15+16])
...
s[31] = reduce(s[31], s[31+16]) // effectively reading outside the allocated 32 elements in shared mem
```

cuda-memcheck is sketchy, I have only managed to detect out-of-bounds issue when I've had really large errors in the addresses.

The cooperative groups are quite recent (9.0), but they are supported on compute capability 3.0+-cards. It probably works on Windows if I set the card to compute mode but I rather not. I think I will try it out if I get a machine at work.

**Comment by m-pilia** [comment]
> What types do you have problems with? Could be an alignment issue.

It just does not compile. Let `T` be `float3`, and let's say  I qualify the shared memory as `volatile` at line 59 (would not do that, but it is a minimal example). Then the compilation fails complainig that `op(float3, float3)` cannot accept an argument list `(volatile float3, volatile float3)`. If I provide an overload `op(volatile float3, volatile float3)`, the next thing to break is the assignment. If I cast to `(T)` then it says that `volatile float3` has no implicit constructor, and so on. I guess I can fix all these issues till it compiles, but the code would become a mess in the process.

> You'll only want to process half the range, but what you're doing now is: [...]

Ok, now I see that, thanks. Yesterday evening I was quite burnt out. But I think that can happen only with a block size (`num_threads`) strictly lesser than 64, while currently it is 256, that's why the code works fine and I did not observe it. Such block sizes are very unlikely to be of any use in a realistic scenario, however, so I was thinking to statically assert a minimum size anyway. Otherwise, if I am not wrong, it should be rather possible to fix it without adding any branching overhead by imposing a minimum allocation of `48` elements in the shared memory.

Curiously, I saw many examples of reductions with this issue, but I guess nobody usually cares because of the small block size that would be affected.

> The cooperative groups are quite recent [...]

Yeah, I definitely wanna read more on the topic, I have never used them.

**Comment by simeks** [comment]
I checked at the optimized code and you're right, your example results in the GPU keeping the result in the registers until the very end, which makes sense. 

I'm wondering if there's actually any benefit in avoiding the synchronization. `__synchthreads` have a latency of 32-64 clock cycles, load/store in shared memory have a latency of around 30 clock cycles (at least for older cards). So in the end, you'll probably have the same performance in both cases, maybe even worse because in your case you would do 3 load/stores.

Yeah, another thing worth trying is to just reduce the remaining values on the CPU if you have less than say, 16 elements. Not sure the cost of invoking kernels and so on.

Funny thing, if you add the `if (tid < 16)` and so on, the optimizer will store the results in shared memory after each step, haha.

**Comment by m-pilia** [comment]
> I'm wondering if there's actually any benefit in avoiding the synchronization. __synchthreads have a latency of 32-64 clock cycles, load/store in shared memory have a latency of around 30 clock cycles (at least for older cards). So in the end, you'll probably have the same performance in both cases, maybe even worse because in your case you would do 3 load/stores.

I agree, here I'm splitting hairs. It may make it a bit faster, but I am rather sure the difference won't be a game changer. It was mostly my stubbornness at being able to make it work that way too.

> Yeah, another thing worth trying is to just reduce the remaining values on the CPU if you have less than say, 16 elements. Not sure the cost of invoking kernels and so on.

That's a neat idea. The annoying part is that there's a lot of variability between different cards (and CPUs), so I guess it is hard to estimate the benefits in a general way.

> Funny thing, if you add the if (tid < 16) and so on, the optimizer will store the results in shared memory after each step, haha.

Lol, I had a look at the ptx, you're right... So it seems the sync is redundant in that case, but that definitely seems a behaviour we cannot rely on...

**Comment by simeks** [comment]
Haha, I recognize that feeling.

If you feel like implementing something like a generic reduction in STK you're free to. The min/max reduction in gpu_volume.cu could be a good starting point. No stress of course, the current reduction works for what I need.
